pkg/instance/instance.go

```diff
@@ -0,0 +1,588 @@
+	logrus.WithFields(logrus.Fields{}).Info("Listing instances")
+
+	instances := map[string]*rpc.InstanceResponse{}
+
+	err := s.processInstanceList(instances)
+	if err != nil {
+		return nil, err
+	}
+
+	if s.spdkEnabled {
+		err := s.spdkInstanceList(instances)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &rpc.InstanceListResponse{
+		Instances: instances,
+	}, nil
+}
+
+func (s *Server) processInstanceList(instances map[string]*rpc.InstanceResponse) error {
+	pmClient, err := client.NewProcessManagerClient("tcp://"+s.processManagerServiceAddress, nil)
+	if err != nil {
+		return err
+	}
+	defer pmClient.Close()
+
+	processes, err := pmClient.ProcessList()
+	if err != nil {
+		return err
+	}
+	for _, process := range processes {
+		instances[process.Spec.Name] = processResponseToInstanceResponse(process)
+	}
+	return nil
+}
+
+func (s *Server) spdkInstanceList(instances map[string]*rpc.InstanceResponse) error {
+	c, err := spdkclient.NewSPDKClient(s.spdkServiceAddress)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	replicas, err := c.ReplicaList()
+	if err != nil {
+		return err
+	}
+	for _, replica := range replicas {
+		instances[replica.Name] = replicaResponseToInstanceResponse(replica)
+	}
+
+	engines, err := c.EngineList()
+	if err != nil {
+		return err
+	}
+	for _, engine := range engines {
+		instances[engine.Name] = engineResponseToInstanceResponse(engine)
+	}
+	return nil
+}
+
+func (s *Server) InstanceReplace(ctx context.Context, req *rpc.InstanceReplaceRequest) (*rpc.InstanceResponse, error) {
+	logrus.WithFields(logrus.Fields{
+		"name":               req.Spec.Name,
+		"type":               req.Spec.Type,
+		"backendStoreDriver": req.Spec.BackendStoreDriver,
+	}).Info("Replacing instance")
+
+	switch req.Spec.BackendStoreDriver {
+	case rpc.BackendStoreDriver_longhorn:
+		return s.processInstanceReplace(req)
+	case rpc.BackendStoreDriver_spdk:
+		return s.spdkInstanceReplace(req)
+	default:
+		return nil, grpcstatus.Errorf(grpccodes.InvalidArgument, "unknown backend store driver %v", req.Spec.BackendStoreDriver)
+	}
+}
+
+func (s *Server) processInstanceReplace(req *rpc.InstanceReplaceRequest) (*rpc.InstanceResponse, error) {
+	if req.Spec.ProcessSpecific == nil {
+		return nil, fmt.Errorf("process is required for longhorn backend store driver")
+	}
+
+	pmClient, err := client.NewProcessManagerClient("tcp://"+s.processManagerServiceAddress, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer pmClient.Close()
+
+	process, err := pmClient.ProcessReplace(req.Spec.Name,
+		req.Spec.ProcessSpecific.Binary, int(req.Spec.PortCount), req.Spec.ProcessSpecific.Args, req.Spec.PortArgs, req.TerminateSignal)
+	if err != nil {
+		return nil, err
+	}
+
+	return processResponseToInstanceResponse(process), nil
+}
+
+func (s *Server) spdkInstanceReplace(req *rpc.InstanceReplaceRequest) (*rpc.InstanceResponse, error) {
+	return nil, fmt.Errorf("spdk instance replace is not supported")
```

Comment:
Use gRPC error instead. Please review all returned errors in gRPC methods.

Reply:
Addressed. Thank you.